File: InMoto/Models/NavigationModels.swift
import Foundation
import CoreLocation

struct GeocodedWaypoint: Codable, Identifiable {
    let id: String
    let name: String
    let latitude: Double
    let longitude: Double

    var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }

    init(name: String, latitude: Double, longitude: Double) {
        self.id = UUID().uuidString
        self.name = name
        self.latitude = latitude
        self.longitude = longitude
    }
}

struct RouteStep: Codable, Identifiable {
    let id: String
    let instruction: String
    let distanceMeters: Double
    let latitude: Double
    let longitude: Double

    var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }

    init(instruction: String, distanceMeters: Double, coordinate: CLLocationCoordinate2D) {
        self.id = UUID().uuidString
        self.instruction = instruction
        self.distanceMeters = distanceMeters
        self.latitude = coordinate.latitude
        self.longitude = coordinate.longitude
    }
}

struct RouteLeg: Codable, Identifiable {
    let id: String
    let fromName: String
    let toName: String
    let distanceMeters: Double
    let durationSeconds: Double
    let polylinePoints: [[Double]]   // [[lat, lon], …]
    let steps: [RouteStep]

    var polylineCoordinates: [CLLocationCoordinate2D] {
        polylinePoints.compactMap { p in
            guard p.count == 2 else { return nil }
            return CLLocationCoordinate2D(latitude: p[0], longitude: p[1])
        }
    }

    init(fromName: String, toName: String, distanceMeters: Double, durationSeconds: Double,
         polylineCoordinates: [CLLocationCoordinate2D], steps: [RouteStep]) {
        self.id = UUID().uuidString
        self.fromName = fromName
        self.toName = toName
        self.distanceMeters = distanceMeters
        self.durationSeconds = durationSeconds
        self.polylinePoints = polylineCoordinates.map { [$0.latitude, $0.longitude] }
        self.steps = steps
    }
}

struct NavigationRoute: Codable {
    let routeId: String
    let routeName: String
    let waypoints: [GeocodedWaypoint]
    let legs: [RouteLeg]

    var totalDistanceMeters: Double { legs.reduce(0) { $0 + $1.distanceMeters } }
    var totalDurationSeconds: Double { legs.reduce(0) { $0 + $1.durationSeconds } }

    var formattedDistance: String {
        String(format: "%.0f km", totalDistanceMeters / 1000)
    }

    var formattedDuration: String {
        let total = Int(totalDurationSeconds)
        let h = total / 3600
        let m = (total % 3600) / 60
        if h == 0 { return "\(m) min" }
        if m == 0 { return "\(h)h" }
        return "\(h)h \(m)min"
    }
}

File: android/app/src/main/java/com/divito/inmoto/model/NavigationModels.kt
package com.divito.inmoto.model

import kotlinx.serialization.Serializable
import java.util.UUID

/** Coordinata geografica semplice (lat/lon). Convertita in LatLng MapLibre al bordo UI. */
@Serializable
data class GeoPoint(val lat: Double, val lon: Double)

// MARK: - TurnDirection

/** Tipo di manovra. L'icona concreta è scelta nel layer UI (Material icons). */
enum class TurnDirection {
    straight, slightLeft, left, sharpLeft,
    slightRight, right, sharpRight,
    uTurn, roundabout, arrive, depart;

    companion object {
        /** Parsa istruzioni in italiano o inglese (MKDirections / OSRM). */
        fun parse(instruction: String): TurnDirection {
            val t = instruction.lowercase()
            if (t.contains("arriv") || t.contains("destinazione") || t.contains("destination")) return arrive
            if (t.contains("torna indietro") || t.contains("inversione") || t.contains("u-turn")) return uTurn
            if (t.contains("rotonda") || t.contains("roundabout")) return roundabout
            if ((t.contains("legger") || t.contains("slight")) && (t.contains("sinistra") || t.contains("left"))) return slightLeft
            if ((t.contains("legger") || t.contains("slight")) && (t.contains("destra") || t.contains("right"))) return slightRight
            if (t.contains("sinistra") || t.contains("left")) return left
            if (t.contains("destra") || t.contains("right")) return right
            if (t.contains("parte") || t.contains("depart")) return depart
            return straight
        }
    }
}

@Serializable
data class GeocodedWaypoint(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val latitude: Double,
    val longitude: Double,
    val note: String? = null,   // nota/descrizione del nodo (es. dai waypoint GPX)
) {
    val point: GeoPoint get() = GeoPoint(latitude, longitude)
}

@Serializable
data class RouteStep(
    val id: String = UUID.randomUUID().toString(),
    val instruction: String,
    val distanceMeters: Double,
    val latitude: Double,
    val longitude: Double,
) {
    val point: GeoPoint get() = GeoPoint(latitude, longitude)
}

@Serializable
data class RouteLeg(
    val id: String = UUID.randomUUID().toString(),
    val fromName: String,
    val toName: String,
    val distanceMeters: Double,
    val durationSeconds: Double,
    val polylinePoints: List<List<Double>>,   // [[lat, lon], …]
    val steps: List<RouteStep>,
) {
    val polylineCoordinates: List<GeoPoint>
        get() = polylinePoints.mapNotNull { p ->
            if (p.size == 2) GeoPoint(p[0], p[1]) else null
        }

    companion object {
        fun from(
            fromName: String,
            toName: String,
            distanceMeters: Double,
            durationSeconds: Double,
            polyline: List<GeoPoint>,
            steps: List<RouteStep>,
        ) = RouteLeg(
            fromName = fromName,
            toName = toName,
            distanceMeters = distanceMeters,
            durationSeconds = durationSeconds,
            polylinePoints = polyline.map { listOf(it.lat, it.lon) },
            steps = steps,
        )
    }
}

@Serializable
data class NavigationRoute(
    val routeId: String,
    val routeName: String,
    val waypoints: List<GeocodedWaypoint>,
    val legs: List<RouteLeg>,
    val transport: String? = null,   // "moto" | "piedi" (null = moto)
) {
    /** Mezzo del percorso (default moto se non specificato). */
    val transportMode: TransportMode get() = TransportMode.from(transport)

    val totalDistanceMeters: Double get() = legs.sumOf { it.distanceMeters }
    val totalDurationSeconds: Double get() = legs.sumOf { it.durationSeconds }

    val formattedDistance: String get() = "%.0f km".format(totalDistanceMeters / 1000)

    val formattedDuration: String
        get() {
            val total = totalDurationSeconds.toInt()
            val h = total / 3600
            val m = (total % 3600) / 60
            return when {
                h == 0 -> "$m min"
                m == 0 -> "${h}h"
                else -> "${h}h ${m}min"
            }
        }
}

// MARK: - FavoritePlace

@Serializable
data class FavoritePlace(
    val id: String = UUID.randomUUID().toString(),
    var tag: String,       // etichetta (es. "Casa", "Lavoro")
    var address: String,   // indirizzo testuale
)
